### ISSUE TITLE
Model for Mesh data-classifcation problem

### ISSUE THREAD

**Issue by shangguan9191** [pyg-team/pytorch_geometric#1455]
Dear Matt, 

do you know models other than SplineCnn, which deal with Mesh data? 
SplineCNN also deal with the node classification, does that mean any model dealing with node classifcation can be used in Mesh?

**Comment by aluo-x** [comment]
So I have been using pytorch geometric for mesh data recently. With 10k+ vertices but sparse (triangles) connectivity. Not for node classification, but for regression tasks. I have not really looked into how pytorch geometric converts face to edges, but have used the super efficient implementation used in the [Neural Subdivision](https://www.dgp.toronto.edu/projects/neural-subdivision/) paper, since I have to do this step repeatedly due to other changes to the mesh.  

A few empirical observations: 
* ReLU activations really don't seem to work well for mesh data. I suspect to due to sparse connectivity, but [ELU](https://pytorch.org/docs/stable/_modules/torch/nn/modules/activation.html#ELU)/[SiLU (Swish)](https://github.com/fastai/fastai2/blob/master/fastai2/layers.py#L610) work pretty well
* Due to limitations on batch size, BatchNorm seems to work poorly, and InstanceNorm is "too strong" and destroys feature correlation. I've been using LayerNorm (see my implementation for [graphs here](https://github.com/rusty1s/pytorch_geometric/issues/687#issuecomment-663281686)).
* You will probably be better off implementing your own graph conv weight inits, since the defaults in pytorch_geometric may not be quite right. I was really surprised that the initialization matters quite a bit
* Residual connections do help in deep graph convs. The ordering with norm/activation also matters. Using the ones proposed by Kaiming [here](https://arxiv.org/abs/1603.05027) yields better results than the original residual blocks
*   Different graph conv layers produce very different results given the same activation and normalization etc., GCN/GAT/EdgeConv work well
* Sum aggregation is not ideal for irregular meshes. Results were consistently inferior to mean aggregation for regression (maybe for classification it would work better). Unfortunately a pain point in Pytorch Geometric is that the aggregation method is not always exposed... 
* You do not get translation invariance for free, but works like MeshCNN and [FeaStConv](https://pytorch-geometric.readthedocs.io/en/latest/modules/nn.html#torch_geometric.nn.conv.FeaStConv) and Neural Subdivision have had to engineer features to get invariance. Otherwise get ready to do data augmentation or centering of the data
* For certain meshes, the diffusion rate does seem to be high leading to over smoothing. I have been combating this with multi-layer outputs from the network, similar to the "jumping knowledge" paper

Pytorch Geometric is a really great library, and has probably saved me dozens of work days :) 
@rusty1s 
But for pytorch geometric, a few things could be improved:
1. Expose aggregation method for each kind of convolution (GAT w/ mean aggregation maybe?)
2. Allow us to define our own initialization method
3. Working with batches is still kind of odd. Documentation on batches & examples are still lacking. Would be nice to have more examples on converting from batch -> datalist, datalist -> batch, and conversions from the output of networks
4. Memory usage still seems high, not sure how to reduce this

**Comment by rusty1s** [comment]
Thanks for your valuable feedback and observations. I would be very interested in the "Neural Subdivision" face to edge transform, so please feel free to share your code :)

To comment on your points:
1. In general, PyG exposes the aggregation for all operators for which that makes sense. For example, in `GATConv` and `GCNConv`, it does not really make sense to allow any other aggregation than `add` since those operators are defined to work that way.
2. I think this is a great idea. However, PyTorch does not seem to allow this either (which guidelines I follow), so initialization needs to be handled manually. In general, we try to follow the initialization scheme applied in the official implementation, which might not always be the best case for meshes. Please let me know if you have any idea on how to allow this.
3. I feel like documentation of the conversion from data list to batch object is pretty well documented. However, the reverse functionality is indeed not. If you simply want to convert a batched node feature tensor back to a list, one can always use internal PyTorch methods, *e.g.*, [`torch.split`](https://pytorch.org/docs/stable/torch.html?highlight=split#torch.split).
4. For most operators, this is not really possible. For some of them, I have written a tutorial on [memory-efficient aggregations](https://pytorch-geometric.readthedocs.io/en/latest/notes/sparse_tensor.html).

**Comment by aluo-x** [comment]
Looking at GATConv more closely, I agree that since the attention is already normalized it doesn't make sense to do mean aggregation. I should have looked more closely.

For intialization, I have just been copying over the layer definitions, and writing my own. Not very clean.

I propose the following:
1. Allow the definition of the initialization method in the layer initialization (with a sane default, such as xavier/he/normal/uniform). This could be either a string (which I think is sane), or a function object (so user defines optional arguments with [functools.partial](https://towardsdatascience.com/partial-functions-in-python-4441270df1e8), I'm not sure this is a good option)
2. Assuming the input is a string. allow for the definition of gain where applicable

Possible [similar to this](https://github.com/NVlabs/SPADE/blob/0ff661e70131c9b85091d11a66e019c0f2062d4c/models/networks/base_network.py#L28)?

**Comment by rusty1s** [comment]
I see. I need to think about it. For example, the proposed way to re-initialize your models in PyTorch is to use `apply` with a custom `init_weights` function, *e.g.*, like [this one](https://gist.github.com/jojonki/be1e8af97dfa12c983446391c3640b68).

**Comment by aluo-x** [comment]
So more experiments recently indicate, that as long as the normalization is not done in the early layers. InstanceNorm & BatchNorm are superior for mesh regression. BatchNorm however is quite noisy at small (5ish per GPU) batch sizes.
